src/persist-client/src/operators/shard_source.rs

```diff
@@ -0,0 +1,467 @@
+                    None => Antichain::new(),
+                };
+
+                let retired_parts = inflight_parts.drain_filter_swapping(|(upper, _size)| {
+                    PartialOrder::less_equal(&*upper, &flow_control_upper)
+                });
+
+                for (_upper, size_in_bytes) in retired_parts {
+                    inflight_bytes -= size_in_bytes;
+                    trace!(
+                        "shard {} returning {} bytes. total: {}. batch frontier {:?} less_equal to {:?}",
+                        shard_id,
+                        size_in_bytes,
+                        inflight_bytes,
+                        _upper,
+                        flow_control_upper,
+                    );
+                }
+            }
+        }
+    });
+
+    (descs_stream, shutdown_button)
+}
+
+pub(crate) fn shard_source_fetch<K, V, T, D, G>(
+    descs: &Stream<G, (usize, SerdeLeasedBatchPart)>,
+    name: &str,
+    clients: Arc<Mutex<PersistClientCache>>,
+    location: PersistLocation,
+    shard_id: ShardId,
+) -> (
+    Stream<G, FetchedPart<K, V, T, D>>,
+    Stream<G, SerdeLeasedBatchPart>,
+    ShutdownButton<()>,
+)
+where
+    K: Debug + Codec,
+    V: Debug + Codec,
+    T: Timestamp + Lattice + Codec64,
+    D: Semigroup + Codec64 + Send + Sync,
+    G: Scope<Timestamp = T>,
+{
+    let mut builder =
+        AsyncOperatorBuilder::new(format!("shard_source_fetch({})", name), descs.scope());
+    let mut descs_input = builder.new_input(
+        descs,
+        Exchange::new(|&(i, _): &(usize, _)| u64::cast_from(i)),
+    );
+    let (mut fetched_output, fetched_stream) = builder.new_output();
+    let (mut tokens_output, tokens_stream) = builder.new_output();
+
+    let shutdown_button = builder.build(move |_capabilities| async move {
+        let fetcher = {
+            let mut clients = clients.lock().await;
+
+            let client = clients
+                .open(location.clone())
+                .await
+                .expect("location should be valid");
+
+            // Unlock the client cache before we do any async work.
+            std::mem::drop(clients);
+
+            client.create_batch_fetcher::<K, V, T, D>(shard_id).await
+        };
+
+        let mut buffer = Vec::new();
+
+        while let Some(event) = descs_input.next().await {
+            if let Event::Data(cap, data) = event {
+                // `LeasedBatchPart`es cannot be dropped at this point w/o
+                // panicking, so swap them to an owned version.
+                data.swap(&mut buffer);
+
+                for (_idx, part) in buffer.drain(..) {
+                    let (token, fetched) = fetcher
+                        .fetch_leased_part(fetcher.leased_part_from_exchangeable(part))
+                        .await;
+                    let fetched = fetched.expect("shard_id should match across all workers");
```

Comment:
We were previously testing the yield function after fetching each part but it looks like now we will fetch an entire slice of times in memory before attempting to yield. Should we move yielding in here?

Reply:
> so previously if we had 10GB worth of parts we would yield before fetching all of them and so if there was a COUNT(*) downstream we could process all 10GB of them with much less memory. Maybe I haven't understood the code correctly, but isn't it the case that all 10GB will be loaded here before we can ever try to slice it up in smaller timely batches?

yeah that's how it behaved before, but like I said in my previous response here, I think it's fine? in practice, we won't wait until they're fetched. one part will arrive first, and then this operator will hit an await, yield to timely, and let the next operator run. it's possible that we get a bunch of awaits in a row that all are ready, but unlikely. IMO, it's a better contract for this operator to be allowed to fetch things as fast as it can and do the backpressure in the operator that emits the desc (as the current flow control stuff does)

(now that I type that out, it does strike me that it assumes an even distribution of parts between workers. perhaps we also (instead?) want to plumb the flow_control_input here and give each fetch worker a budget of `flow_control_max_inflight_bytes/scope.peers()`?)

Reply:
mulled this over and realized I could say the above much more simply. proposal: going forward, we backpressure at the granularity of parts using flow_control but "backpressure" within a part with yield_fn